File: Python/Functions/SpectralRepresentationMethod.py
import numpy as np
###################################################################################################################################
#Implementation of the Spectral Representation Method, Shinozuka & Deodatis (1991)
###################################################################################################################################
def SpectralRepresentationMethod(S, w, t):
    Nt = len(t)
    Nw = len(w)
    dw = w[1] - w[0]
    
    x_shnzk = np.zeros(Nt)
    
    for w_n in range(Nw):
        if w[w_n] == 0:
            A = 0
        else:
            A = np.sqrt(2 * S(w[w_n]) * dw)
        
        phi = np.random.rand() * (2 * np.pi)
        x_shnzk += np.sqrt(2) * A * np.cos(w[w_n] * t + phi)
    
    return x_shnzk

File: MATLAB/Functions/SpectralRepresentationMethod.m
function x_shnzk = SpectralRepresentationMethod(S, w, t)
%Function sampling a single stochastic process for a given frequency and
%time discretization as well as an anonymous Power Spectral Density
%function SRM, based on Shinozuka & Deodatis 1991.

tic_shnzk = tic;

Nt = length(t);
Nw = length(w);
dw = w(2)-w(1);

x_shnzk = zeros(1,Nt);

for w_n = 1:Nw
    
    if w(w_n) == 0
        A = 0;
    else
        A = sqrt(2*S(w(w_n))*dw);         
    end
    
    phi = rand*(2*pi);
    x_shnzk(1,:) = x_shnzk(1,:) + sqrt(2) * A * cos(w(w_n) * t + phi);
end

toc_shnzk = toc(tic_shnzk);
disp(['Shnzk elapsed time: ' num2str(toc_shnzk) 's'])

end